USWDS - Validation: Allow textareas to use validation (Enhancement)

danbrady opened the pull request:
# Summary

**Textarea components can use validation** Now you can use validation on textareas, just like you do on inputs.


## Breaking change

This is not a breaking change.

## Related issue

Closes #4702


## Problem statement

Currently, validation can only be used on `input` elements. This solves most validation needs. However, `textarea` is a form element in which validation might be desired (e.g. not empty, minimal number of characters or words.)


## Solution

1. Current selector for validation is `input[data-validation-element]`
2. Updated validation logic to include `textarea[data-validation-element]`



## Testing and review

_Share recommended methods for reviewing this change._
<!--
1. Describe the tests that you ran to verify your changes,
2. Provide instructions to reproduce these tests, and
3. Clarify the type of feedback you are looking for at this phase.
-->

1. Added `textarea` to tests

## Dependency updates
None.

amyleadem commented:
Thanks for this submission! I have tagged this for review and we will take a look at our next available opportunity. If you have any questions in the meantime, please let us know!

danbrady commented:
Good question, @mejiaj. Is it fair to think character count is validation for a **single, specific requirement**? (I.e. cannot be more than N characters.) From the [docs](https://designsystem.digital.gov/components/character-count/):

> Character count helps users know how much text they can enter when there is a limit on the number of characters.

(Admittedly, this is probably the most useful validation for a `textarea`.)

However, allowing `textarea` to use the validation package allows it to provide visual feedback for multiple validation requirements that a `textarea` could have. For example: must have a _minimal_ number of characters/words.

The use cases are likely rare, but are not non-existent.

mejiaj commented on packages/usa-validation/src/usa-validation.twig:
```diff
@@ -1,17 +1,33 @@
+		<label class="usa-label" for="comment">Comment</label>
+		<textarea class="usa-textarea" id="comment" name="comment" data-validate-empty="\S" data-validation-element="validate-comment"></textarea>
+		<input class="usa-button" type="submit" value="Submit comment">
+	</fieldset>
```
Can we separate these tests, one for input and another for text area?

mejiaj commented on packages/usa-validation/src/test/validator.spec.js:
Should we add a test that mentions initialization on textarea?

danbrady replied:
Agreed. I'll separate `textarea` into separate story, test, etc. and add this to it instead of mixing it up with `input`. I'll update the PR...

mejiaj reviewed (changes requested):
Looks good as far as extending the functionality to textarea. Just some comments below.

danbrady commented:
Ok, @mejiaj, I split textarea out. A few observations when looking at the tests on the original `input` validation:

https://github.com/uswds/uswds/blob/18e648028a624f387d7999fd6f2c287c56f021ad/packages/usa-validation/src/test/validator.spec.js#L10

I don't see an element with that `data-checklist-label` selector in any template?

https://github.com/uswds/uswds/blob/18e648028a624f387d7999fd6f2c287c56f021ad/packages/usa-validation/src/test/validator.spec.js#L47-L53

`queryElement` isn't a valid method (perhaps `querySelector` was intended?), however execution never makes it to that point to throw an error because the `validators` object is always empty...

https://github.com/uswds/uswds/blob/18e648028a624f387d7999fd6f2c287c56f021ad/packages/usa-validation/src/test/validator.spec.js#L13

https://github.com/uswds/uswds/blob/18e648028a624f387d7999fd6f2c287c56f021ad/packages/usa-validation/src/test/validator.spec.js#L37-L39

(It's looking for the validators _within_ the `input` and not the validation checklist.)

Amazingly, all the test pass! You can even change all the asserts to anything and it still passes. 😄 

I addressed all these in the `textarea` tests. Please let me know if I'm on the right track with them? And if I am, should I update the tests for `input` as part of this PR? Thanks!

mejiaj commented:
Thanks for making these changes and apologies for the late reply.
 
### Data checklist label

> I don't see an element with that `data-checklist-label` selector in any template?

Looks like a regression. This was introduced in #4914. Looks like in rework it was removed in commit [c284450](https://github.com/uswds/uswds/pull/4914/commits/c284450b7f8aa0f4cffc872fb8f809c48f8f39b7), but the test wasn't updated. Thanks for catching.

### Unit tests

#### Validation list items

> `queryElement` isn't a valid method (perhaps `querySelector` was intended?), however execution never makes it to that point to throw an error because the `validators` object is always empty...

Yes, that looks incorrect.

#### Incorrect validators

> https://github.com/uswds/uswds/blob/18e648028a624f387d7999fd6f2c287c56f021ad/packages/usa-validation/src/test/validator.spec.js#L13
> 
> https://github.com/uswds/uswds/blob/18e648028a624f387d7999fd6f2c287c56f021ad/packages/usa-validation/src/test/validator.spec.js#L37-L39
> 
> (It's looking for the validators _within_ the `input` and not the validation checklist.)

This should fix this. Validators are set in list items and should be checked there.

It's also a good callout for improving our testing, thank you.

mejiaj commented:
## Todo

- [ ] Remove unused data attributes
- [ ] Remove unused references in unit tests

mahoneycm commented on packages/usa-validation/src/index.js:
```diff
@@ -3,7 +3,8 @@ const validate = require("../../uswds-core/src/js/utils/validate-input");
 const selectOrMatches = require("../../uswds-core/src/js/utils/select-or-matches");
 
-const VALIDATE_INPUT = "input[data-validation-element]";
+const VALIDATE_INPUT =
+  "input[data-validation-element],textarea[data-validation-element]";
```
## Validate input selector

Curious if there is a slightly cleaner way to define this selector. Probably not a blocker because this works well but may be room for a future enhancement in the future 🤔

mahoneycm commented on packages/usa-validation/src/test/template-textarea.html:
## Validation Alert: Defined width (future enhancement)

Currently, the validation alert matches the size of the input it's tied to. With text area, users can resize the text box using the corner of the input and it causes the alert to grow with it.

I'm curious if it'd be a good idea to set a max width on the alert or to limit the text area to only extend vertically. Flagging for consideration!

[Kapture 2024-02-05 at 17.21.47.webm](https://github.com/uswds/uswds/assets/25211650/bab1e004-60f7-4f6c-a2e8-2f3acff75548)

amyleadem replied:
I don't mind this behavior, actually, but that might just be personal preference. Happy to discuss this as a possible future change if desired. I don't think it needs to be a blocker for this PR.

mahoneycm reviewed (commented):
Component, stories, and tests look good! Left two comments for potential future enhancements for the team.

CC: @amyleadem @mejiaj

amyleadem reviewed (approved):
Looks good to me!

Here is what I tested:
- [x] Confirm that the validation component works with `textarea` elements
- [x] Confirm that a new story for the `textarea` variant exists and works as expected
- [x] Confirm unit tests work as expected

thisisdano reviewed (approved):
LGTM

aduth commented:
It looks like the changes here may have introduced an accessibility build failure which is causing other pull requests to fail.

```
  1) Components/Validation
       Textarea Validation:
     Detected the following accessibility violations!

     1. label (Form elements must have labels)

        For more info, visit https://dequeuniversity.com/rules/axe/4.8/label?application=axeAPI.

        Check these nodes:

        - html: <textarea class="usa-textarea" id="comment" name="comment" data-validate-empty="\S" data-validation-element="validate-comment"></textarea>
          summary: Fix any of the following:
                     Form element does not have an implicit (wrapped) <label>
                     Form element does not have an explicit <label>
                     aria-label attribute does not exist or is empty
                     aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty
                     Element has no title attribute
                     Element has no placeholder attribute
                     Element's default semantics were not overridden with role="none" or role="presentation"
```
